### ISSUE TITLE
EasySIMBL stopped working on 10.10.4 b4

### ISSUE THREAD

**Issue by antons** [norio-nomura/EasySIMBL#25]
Just restarted to 10.10.4 b4 and noticed that EasySIMBL stopped working. All checkboxes are enabled, but the plugins are not injected into apps.

**Comment by norio-nomura** [comment]
It seems OS X 10.10.4 b4 (14E26a) has restricted API that EasySIMBL (and original SIML) depend on  with some undocumented conditions.

> 2015/05/27 12:57:34.421 com.github.norio-nomura.SIMBL-Agent[41695]: AppleEvents: Send port for process has no send right, port=( port:32047/0x7d2f rcv:1,send:0,d:0 limit:5) (findOrCreate()/AEMachUtils.cp #526) com.apple.main-thread

**Comment by d235j** [comment]
This warning appeared on previous 10.10 versions. So I doubt this is why it fails.

**Comment by norio-nomura** [comment]
I have been taught that the log is appearing also on OS X 10.10.3.
It seems silently failed injecting on OS X 10.10.4 (14E26a)
EasySIMBL has ever stopped working while OS X 10.9 Preview as #4, but it works again until final release.

**Comment by d235j** [comment]
This is a weird one. TotalFinder's injector (source at https://github.com/binaryage/totalfinder-osax) still works, as does FinderPop's (finderpop.com), Default Folder X, Bartender.

Apple also changed other major stuff in this 10.10.4 beta — resolved was removed and replaced with the older mDNSResponder.

**Comment by d235j** [comment]
If I make a standalone AppleScript application that sends the appropriate Apple Event to the desired application, the bundles get loaded. Therefore this is a bug in the EasySIMBL Agent.

**Comment by norio-nomura** [comment]
@d235j Thanks. I'll investigate.

**Comment by norio-nomura** [comment]
I tested injecting EasySIMBL.osax by using AppleScript as following:
```AppleScript
tell application "Safari" to «event ESIMload»
```
got same errors on both OS X 10.10.3 and OS X 10.10.4 as following:
```
tell application "Safari"
	«event ESIMload»
		--> error number -1708
	«event ascrgdut»
		--> error number -1708
	«event ESIMload»
		--> error number -10004
end tell
tell current application
	«event ESIMload»
end tell
```
It works on OS X 10.10.3 (14D131), and not on OS X 10.10.4 (14E26a) :confused:

**Comment by d235j** [comment]
I also had to change Context: User to Context: Process in the OSAX Info.plist.

Now it requires a password when running the Apple Event but at least it loads.

**Comment by norio-nomura** [comment]
@d235j Thank you for explanation,
After editing Info.plist of EasySIMBL.osax 0.10.10 (bundled into EasySIMBL-1.7.1) and re-codesign it, I got works by AppleScript.

**Comment by ylluminate** [comment]
So glad to see this guys, thanks for your effort on this!  This was a real headache for me today.

**Comment by orbitly** [comment]
Any word on when the fix will be released?

**Comment by norio-nomura** [comment]
Referring information about TotalFinder, I have tried fix. But I have not yet success EasySIMBL working on OS X 10.10.4 (14E26a). :disappointed:

**Comment by ylluminate** [comment]
Have we asked the BinaryAge guys for input / help? @darwin @sdsykes

**Comment by sdsykes** [comment]
Don't know why the problem has arisen. You might try using the AESendMessage API rather than SBApplication to send the events.

**Comment by darwin** [comment]
ScriptingAdditions are fragile.

AFAIK, OSAX bundles should reside in `/Library/ScriptingAdditions` (probably installed by root/admin) to be first-class citizens with full rights to be injected in all bundles in /Applications. Having them in `~/Library/ScriptingAdditions` sometimes works, but I had issues that sometimes the system prevents apple events because of "no send right". I don't remember the details, but I would assume system-level OSAXes can be injected into anything. And user-level OSAXes can be injected only into user-level processes. The exact rules are unknown to me but it is possible Apple made some changes of this mechanism in 10.10.4 (14E26a).

Also one thing which burned me during development is that OSAX bundles are being actively scanned for on the whole filesystem using Spotlight or something like that. So if you have some broken OSAX lying somewhere then scripting additions could pick it up and try inject it instead of good OSAX in Library/ScriptingAdditions folders. So for example installer should not unpack OSAX somewhere into temp folder and copy it later to the destination. This could confuse the system in rare cases.

In TotalFinder I had a lot of problems related to failed OSAX injections. I decided not to use hihg-level applescript, but send events using AESendMessage, it boils down to:
https://gist.github.com/darwin/72be854b54c590d05958

Some dev notes:
https://github.com/binaryage/totalfinder-osax/blob/280930f6e3dc96602d31b281a61104da773f2fed/TotalFinderInjector.m#L21-L65

**Comment by norio-nomura** [comment]
@sdsykes @darwin Thanks for informations,
I will try using AESendMessage as first step.

**Comment by d235j** [comment]
@darwin, @norio-nomura It seems that putting the OSAX in ~/Library/Scripting Additions is not the problem since using an AppleScript Editor created application to send the Apple Event still works. This is probably due to how SBApplication is being used.

**Comment by norio-nomura** [comment]
I have tested using `AESendMessage` without `SBApplication`. It works on OS X 10.10.3 and does not work on OS X 10.10.4 (14E26a).
Is it needed to change bundle structures for placing OSAX in `/Library/ScriptingAdditions`? :unamused:

**Comment by antons** [comment]
@norio-nomura In my (limited) testing, that is the only location that worked.

**Comment by antons** [comment]
EasySIMBL doesn’t work on El Capitan too, as far as I can see.

**Comment by orbitly** [comment]
I can report the same, 10.11 (15A178w)

**Comment by w0lfschild** [comment]
Are you sure about that? It's definitely working for me. Granted not all bundles are working with the new apps (cDock, colorfulsidebar). I believe some apps have moved to swift because I can't class dump the dock executable anymore.

**Comment by orbitly** [comment]
Neither Flashlight nor Safari Stand work.

**Comment by w0lfschild** [comment]
That's doesn't mean that simbl isn't working though...

NotificationClear and RadonChrome are both 100% working on 10.11 via easySIMBL.

**Comment by antons** [comment]
@w0lfschild You’re right, some plugins work. However, I have an app and a plugin that have not changed since 10.10, and yet they fail to load.

```
09.06.15 16:06:43,511 Gitbox[14462]: Error loading /Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL:  dlopen(/Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL, 262): no suitable image found.  Did find:
	/Users/Anton/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL: open() failed with errno=24
```

**Comment by d235j** [comment]
From https://developer.apple.com/library/prerelease/mac/releasenotes/MacOSX/WhatsNewInOSX/Articles/MacOSX10_11.html#//apple_ref/doc/uid/TP40016227-SW1 :
> System Integrity Protection
> A new security policy that applies to every running process, including privileged code and code that runs out of the sandbox. The policy extends additional protections to components on disk and at run-time, only allowing system binaries to be modified by the system installer and software updates. Code injection and runtime attachments are no longer permitted.

Supposedly this can be turned off from the Recovery partition

**Comment by antons** [comment]
@d235j It looks like that not doesn’t apply to SIMBL. [One of my plugins](https://github.com/antons/Center-Key-Window) is injected into most apps without any issues.

**Comment by antons** [comment]
@d235j @norio-nomura Checking back, I realize that that plugin was injected into my apps because I placed a copy of `SIMBL.osax` into `/Library/ScriptingAdditions`. After restarting machine without it, none of my plugins work. So the issues in 10.10.4 and 10.11 appear to be related.

**Comment by d235j** [comment]
@norio-nomura:

In my testing, it apparently tries to load when in /Library/ScriptingAdditions. However, I don't have a signing key so I can't test it further, as I get this error:
```
2015-06-13 20:31:41.930 Safari[30755:2250532] Error loading /Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL:  dlopen(/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL, 262): no suitable image found.  Did find:
	/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL: code signature invalid for '/Library/ScriptingAdditions/EasySIMBL.osax/Contents/MacOS/EasySIMBL'
Safari: OpenScripting.framework - can't find entry point InjectEventHandler in scripting addition /Library/ScriptingAdditions/EasySIMBL.osax.
```

To make testing easier, open Terminal and run these two commands:

```
export AEDebugSends=1
export AEDebugReceives=1
```
Then run Safari from Terminal, as follows:
```
/Applications/Safari.app/Contents/MacOS/Safari
```
And you should see all Apple Event activity.

**Comment by norio-nomura** [comment]
Thanks all for testing on OS X 10.10.4 b4.

I should write about my EasySIMBL usage here.
The past two years, I have not used EasySIMBL on my Mac except on testing it.
I have no reason to using SIMBL plugins as I needed on my forking EasySIMBL from original SIMBL.
I don't have time for fixing this issue in the near future.

Thank you for your understanding.

**Comment by ryanp-me** [comment]
Perhaps this is discussed somewhere else, but is there a reason SIMBL must be injected using ScriptingAdditions? I've always used the `DYLD_INSERT_LIBRARIES` environment variable. (Although, I guess if an app set this in it's plist, it could override the SIMBL one. And this won't fix 10.11 support with SIP enabled.)

**Comment by norio-nomura** [comment]
When I forked EasySIMBL, another implementation was exists that using another injecting method.
On that time, injecting dynamically was needed for some system processes(e.g. Finder, Dock
I don't know `DYLD_INSERT_LIBRARIES` can be used for such processes.

**Comment by norio-nomura** [comment]
Notes for someone who creating new.

- Placing plugins under /Library/* have a possibility of freeing from injecting it to containers of sandboxed apps.
- Resume apps on login is happened before launching SIMBL Agent.
- If placing .osax and/or plugins under /System/* while SIP disabled, cann't we load it after SIP enabled on OS X 10.11?

**Comment by therealmarv** [comment]
related I think because it was also previously mentioned here: http://blog.binaryage.com/el-capitan-update/ Total Finder will stop supporting OS X 10.11+
Still hope we will find a way :)

**Comment by norio-nomura** [comment]
I noticed that @d235j's test is similar to original SIMBL.
So I tested [SafariStand 8.0.211](http://hetima.com/safaristand/) with [original SIMBL-0.9.9](http://www.culater.net/software/SIMBL/SIMBL.php) on OS X 10.10.4 beta 4, and confirmed it works.

**Comment by d235j** [comment]
In order to fix this we'd probably just need to fix up the plists, and make it so that EasySIMBL puts its ScriptingAddition in /Library and not ~/Library. If I have time I'll take a look.

**Comment by w0lfschild** [comment]
Yup @d235j and it would seem as if this [security measure](https://support.apple.com/kb/HT204942) is present now in 10.8.5, 10.9.5 and 10.10 versions prior to 10.10.4.

I know for sure it's in the update for 10.9.5 and 10.10.4 because I've got a bunch of emails about it already.

So EasySIMBL is basically dead... RIP :-1:

**Comment by d235j** [comment]
@w0lfschild: what entry are you referring to in the list? I can't seem to see it, at least at first glance.

As I said, since it works from /Library, this would probably be the best way to go about it. I don't think using /System/Library is preferable — at least for now.

**Comment by w0lfschild** [comment]
@d235j Not sure what entry it is, I just know that bunch of people emailed me saying they installed the new security update 2015-005 1.0 and now cDock (aka EasySIMBL) is not working.

**Comment by norio-nomura** [comment]
> So EasySIMBL is basically dead... RIP :-1:

I agree.
I think we can start over from SIMBL-0.9.9
https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426

**Comment by norio-nomura** [comment]
I checked EasySIMBL-1.7.1 on OS X versions updated today.
- On OS X 10.10.4 (14E46), it does not work.
- On OS X 10.9.5 (13F1096), it does not work.
- On OS X 10.8.5 (12F2542), it works.

**Comment by d235j** [comment]
@norio-nomura:
I think it may be better to leave it as-is, but move the ScriptingAddition to /Library (and add logic to request privilege elevation in order to do that when turning on EasySIMBL). In my tests it works from /Library.

**Comment by norio-nomura** [comment]
@d235j: Placing at /Library/ScriptingAdditions does not work on OS X 10.11 (15A204h) https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426
It may be not working on future beta.
But if I might do something about SIMBL, I might start over from SIMBL-0.9.9.
I will not change EasySIMBL.

**Comment by d235j** [comment]
@norio-nomura: did you change the Info.plist to have Context: Process instead of User? I needed to do that to make it work on 10.10.4. On 10.11, System Integrity Protection must be off for it to inject system apps; I haven't tested it yet but I will tomorrow when I have access to a machine running 10.11.
(I also made ThreadSafe: False instead of True in the plist, for safety.)

**Comment by norio-nomura** [comment]
@d235j No, I didn't change Info.plist both of testing on 10.10.4 b4 and 10.11.

**Comment by d235j** [comment]
@norio-nomura can you try that together with /Library/ScriptingAdditions? (You also have to remove EasySIMBL.osax from ~/Library/ScriptingAdditions or it will attempt to use that copy and fail.) I could not get it to work at all until I made this change, even with manual injection with an AppleScript, but it seems to work when set up that way on 10.10.4 at least.

**Comment by norio-nomura** [comment]
@d235j I'm using original SIMBL-0.9.9 on my testing. Because EasySIMBL has some unnecessary functions for placing at /Library/ScriptingAdditions.

**Comment by d235j** [comment]
@norio-nomura: I also had to delete EasySIMBL.osax from ~/Library/ScriptingAdditions.
I'll summarize what it took to make it work:

1. Modify the Info.plist as described above.
2. Install and turn on EasySIMBL as usual.
3. Quit the EasySIMBL app.
4. Copy EasySIMBL.osax to /Library/ScriptingAdditions.
5. Delete EasySIMBL from ~/Library/ScriptingAdditions.
6. Observe that it functions. If something puts it back in ~/Library/ScriptingAdditions, it will stop functioning.

the aforementioned AEDebugSends and AEDebugReceives environment variables are useful for testing.

**Comment by norio-nomura** [comment]
@d235j `SIMBL Agent.app` copies `EasySIMBL.osax` to `~/Library/ScriptingAdditions` before sending apple events. If target app is sandboxed, it also copies .osax and plugins into their container.
Placing at `/Library/ScriptingAdditions` make those functions are unnecessary.

**Comment by d235j** [comment]
@norio-nomura Yeah, understood. I don't know if it still works if the target is sandboxed, though.
Having the app to be able to turn SIMBLE plugins on and off is handy, by the way.

**Comment by sambuev** [comment]
Can you update EasySimbl to newer version where everything is works under 10.10.4? please

**Comment by inket** [comment]
@sambuev It's pointless. It's a lot of work for no benefit (besides the plugins management GUI) because SIMBL already works.

And it's even less useful because 10.11 breaks both EasySIMBL and SIMBL without huge workarounds.

**Comment by norio-nomura** [comment]
I wrote a note about injection mechanism of EasySIMBL extended from SIMBL-0.9.9
https://gist.github.com/norio-nomura/f4346e9cb9db0a411779

**Comment by violinday** [comment]
Not working for me as well.

**Comment by norio-nomura** [comment]
[How to transition from EasySIMBL to SIMBL](https://gist.github.com/hetima/51b71091aedf622f61f7) by hetima

**Comment by gkarag** [comment]
@norio-nomura Hello and thanks for your support. I followed the advice of hetima and now I have SafariStand working. One thing has caught my attention though, in the console I get these:
```
04.07.2015 02:30:49.992 Safari[539]: Performance: Please update this scripting addition to supply a value for ThreadSafe for each event handler: "/Library/ScriptingAdditions/SIMBL.osax"
```
My question is should I be concerned about that and is there a rather easy way to fix it?

**Comment by norio-nomura** [comment]
@gkarag The message means SIMBL(and plugins loaded by that) will be initialized on main thread that drives UI. If those behavior affect you, you will see some UI freeze or beachball cursor on loading plugins. I don't know such plugins are exists or not.

**Comment by Sufl** [comment]
I have just created an account in order to say THANK YOU to norio-nomura for her information and effort above on How to transition from EasySIMBL to SIMBL by hetima because it allowed me to get the colourful sidebar back.

I was distraught when it disappeared with Yosemite 10.10.4 and I read the various threads on this with desperation because it was awful to lose my beautiful sidebar when EasySIMBL no longer worked.  Now thankfully I have it back and I want to say thank you very much for the work and effort that it has taken so that people can enjoy the lovely colourful sidebar again.  I am now very happy again.

Cheers.

**Comment by thekrynn** [comment]
SIMBL stopped working for me as well on 10.10.4. Tried the EasySIMBL to SIMBL, and no luck, although maybe it has to do with the fact that I am trying to use the TheLostFinderNewWindow, which I use on all my machines. Here's what I'm seeing when my machine reboots:

Jul 12 02:06:49 macbookpro2 com.apple.xpc.launchd[1] (net.culater.SIMBL.Agent): Please switch away from OnDemand to KeepAlive.

Jul 12 02:06:49 macbookpro2 com.apple.xpc.launchd[1] (com.apple.xpc.launchd.user.501.100005.Aqua): Could not import service from caller: caller = otherbsd.238, service = com.github.norio-nomura.SIMBL-Agent, error = 119: Service is disabled

Jul 12 02:06:49 macbookpro2.w.21.intranet otherbsd[238]: Could not submit LoginItem job com.github.norio-nomura.SIMBL-Agent: 119: Service is disabled

**Comment by norio-nomura** [comment]
@thekrynn For injecting plugins into Finder using SIMBL-0.9.9, you need doing additional steps.

## Injecting SIMBL to Finder by using Apple Script

### Basic instructions:

1. Open `Script Editor.app`
2. Open new script by using menu `File` > `New` or hitting `⌘N`
3. Paste `tell application "Finder" to inject SIMBL into Snow Leopard` to the script.
4. Run the script by using menu `Script` > `Run` or hitting `⌘R`
<img width="613" alt="screenshot 2015-07-12 16 00 13" src="https://cloud.githubusercontent.com/assets/33430/8636797/2038f944-28af-11e5-8b3f-5ea812d870b2.png">

### If your plugin works with above steps, let’s make your injector app with following steps:

5. Using menu `File` > `Export…`
6. Input `inject SIMBL to Finder` into `Export As:` field
7. Select `Application` on `File Format:` field
8. Click `Save`

<img width="458" alt="screenshot 2015-07-12 15 52 07" src="https://cloud.githubusercontent.com/assets/33430/8636770/0ac22dde-28ae-11e5-8791-aae9c8605f5a.png">

Now you can register your `inject SIMBL to Finder.app` to launch on your login as following:
<img width="780" alt="screenshot 2015-07-12 15 55 02" src="https://cloud.githubusercontent.com/assets/33430/8636780/6d1814f8-28ae-11e5-8b8c-220f5f4c8b8d.png">

This method may work with another plugins for another apps.

**Comment by tkoyn** [comment]
thekrynn, Did you find any way to workaround this to get TheLostFinderNewWindow to work again?

norio-nomura, Are you planning on developing a replacement or update for EasySIMBL?

Thanks

**Comment by thekrynn** [comment]
<img width="812" alt="screen shot 2015-07-19 at 12 07 12 am" src="https://cloud.githubusercontent.com/assets/956099/8764483/2fe5730e-2daa-11e5-9d82-d433c74f64a0.png">
<img width="532" alt="screen shot 2015-07-19 at 12 07 02 am" src="https://cloud.githubusercontent.com/assets/956099/8764484/2ff3b0a4-2daa-11e5-96f6-80619b918fb2.png">

Was working perfectly with 10.10.4 (3 different machines) using the applescript app approach (even with manually double clicking). However, this error occurs on 10.11 DP3:

Finder got an error: A privilege violation occurred. (-10004)

**Comment by thekrynn** [comment]
The following article works just fine with TheLostFinderNewWindow in 10.11. Rebooting in recovery mode and disabling the new security option does the trick. While it's supposed to be for our protection, I hate to say that not being able to open a window with a traditional double click is too much of an inconvenience.

http://blog.binaryage.com/el-capitan-update/

**Comment by norio-nomura** [comment]
@thekrynn I wrote about OS X 10.11 (15A204h) on https://github.com/norio-nomura/EasySIMBL/issues/26#issuecomment-117028426. I don't test yet on newer build.

**Comment by yuchenlin** [comment]
It doesn't work on 10.10.5  as well.

**Comment by tkoyn** [comment]
Yuchelin, Exactly what and how did you test? The post 2 posts above yours reported a way to get Lost Finder New Window to work in 10.11.

**Comment by ashishb** [comment]
Working on 10.10.5 using [norio-nomura's apple script](https://github.com/norio-nomura/EasySIMBL/issues/25#issuecomment-120697440)

**Comment by Gunny123** [comment]
So TypeStatus for Mac 1.0 does not work any longer? Or do I use another method mentioned above?

**Comment by vhristev** [comment]
I make it work on 10.10.5 but its buggy ....
1. Open spotlight everything is working with FlashLight but after couple of searches it crash. 
2. Open spotlight again and there is no FlashLight plugins
3. Initiate again the norio-nomura apple script 
4. Everything is back to normal 
5. Open spotlight and start searching .... it crash randomly so its kind of random loop if its going to work on not
